File: src/main/java/io/github/mdsimmo/bomberman/commands/game/RunStart.java
package io.github.mdsimmo.bomberman.commands.game;

import io.github.mdsimmo.bomberman.commands.Cmd;
import io.github.mdsimmo.bomberman.commands.GameCommand;
import io.github.mdsimmo.bomberman.events.BmRunStartCountDownIntent;
import io.github.mdsimmo.bomberman.game.Game;
import io.github.mdsimmo.bomberman.messaging.Message;
import io.github.mdsimmo.bomberman.messaging.Text;
import io.github.mdsimmo.bomberman.utils.NumberUtils;
import org.bukkit.command.CommandSender;

import java.util.List;

public class RunStart extends GameCommand {

	public RunStart(Cmd parent) {
		super(parent);
	}

	@Override
	public Message name() {
		return context(Text.START_NAME).format();
	}

	@Override
	public List<String> gameOptions(List<String> args) {
		return null;
	}
	
	@Override
	public boolean gameRun(CommandSender sender, List<String> args, Game game) {
		if (args.size() > 1)
			return false;

		// get the delay
		var delay = 3;
		if (args.size() == 1) {
			var parsed = NumberUtils.tryParseInt(args.get(0));
			if (parsed.isPresent()) {
				delay = parsed.getAsInt();
				if (delay < 0) {
					Text.INVALID_NUMBER.with("number", delay).sendTo(sender);
					return true;
				}
			} else {
				return false;
			}
		}

		var e = BmRunStartCountDownIntent.startGame(game, delay);
		if (e.isCancelled()) {
			e.getCancelledReason().ifPresentOrElse(
					reason -> reason.sendTo(sender),
					() -> Text.GAME_START_CANCELLED.with("game", game).sendTo(sender)
			);
		} else {
			Text.GAME_START_SUCCESS.with("game", game).sendTo(sender);
		}
		return true;
	}

	@Override
	public Permission permission() {
		return Permission.GAME_OPERATE;
	}

	@Override
	public Message extra() {
		return context(Text.START_EXTRA).format();
	}

	@Override
	public Message example() {
		return context(Text.START_EXAMPLE).format();
	}

	@Override
	public Message description() {
		return context(Text.START_DESCRIPTION).format();
	}

	@Override
	public Message usage() {
		return context(Text.START_USAGE).format();
	}
}

File: src/main/java/io/github/mdsimmo/bomberman/commands/game/RunStart.kt
package io.github.mdsimmo.bomberman.commands.game

import io.github.mdsimmo.bomberman.commands.Cmd
import io.github.mdsimmo.bomberman.commands.GameCommand
import io.github.mdsimmo.bomberman.commands.Permission
import io.github.mdsimmo.bomberman.commands.Permissions
import io.github.mdsimmo.bomberman.events.BmRunStartCountDownIntent
import io.github.mdsimmo.bomberman.game.Game
import io.github.mdsimmo.bomberman.messaging.Formattable
import io.github.mdsimmo.bomberman.messaging.Message
import io.github.mdsimmo.bomberman.messaging.Text
import org.bukkit.command.CommandSender

class RunStart(parent: Cmd) : GameCommand(parent) {
    override fun name(): Formattable {
        return Text.START_NAME
    }

    override fun gameOptions(args: List<String>): List<String> {
        return emptyList()
    }

    override fun flags(sender: CommandSender): Set<String> {
        return setOf("d", "o")
    }

    override fun flagExtension(flag: String): Formattable {
        return when (flag) {
            "d" -> Text.START_FLAG_DELAY_EXT
            else -> Message.empty
        }
    }

    override fun flagDescription(flag: String): Formattable {
        return when (flag) {
            "d" -> Text.START_FLAG_DELAY_DESC
            "o" -> Text.START_FLAG_OVERRIDE_DESC
            else -> Message.empty
        }
    }

    override fun gameRun(sender: CommandSender, args: List<String>, flags: Map<String, String>, game: Game): Boolean {
        if (args.isNotEmpty()) {
            return false
        }
        // get the delay
        val delay = when (val s = flags["d"] ?: flags["delay"]) {
            null -> 3
            else -> {
                val i = s.toIntOrNull()
                if ((i == null) || (i < 0)) {
                    Text.INVALID_NUMBER.format(cmdContext().plus("number", s)).sendTo(sender)
                    return true
                } else {
                    i
                }
            }
        }

        val e = BmRunStartCountDownIntent.startGame(game, delay, flags.containsKey("o"))
        if (e.isCancelled) {
            (e.cancelledReason
                    ?: Text.COMMAND_CANCELLED.format(cmdContext()
                            .plus("game", game)))
                .sendTo(sender)
        } else {
            Text.GAME_START_SUCCESS.format(cmdContext().plus("game", game)).sendTo(sender)
        }
        return true
    }

    override fun permission(): Permission {
        return Permissions.START
    }

    override fun extra(): Formattable {
        return Text.START_EXTRA
    }

    override fun example(): Formattable {
        return Text.START_EXAMPLE
    }

    override fun description(): Formattable {
        return Text.START_DESCRIPTION
    }

    override fun usage(): Formattable {
        return Text.START_USAGE
    }
}